### DISCUSSION TITLE
Add transform param to Path, Query, Body class which will transform the field in the desired type and value

### DISCUSSION THREAD

**Discussion by Niraj-Kamdar** [fastapi/fastapi#6689] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * Or, I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI

app = FastAPI()


@app.get("/{animal_name}")
def get_animal(animal_name: str = Path(transform=lambda x: x.lower()):
    return get_animal_from_db(animal_name)
```

### Description

This will be helpful in pre-processing data in desired format. We can do it inside the function but that can obscure some readability. Having it this way will make code more readable.

### Describe alternatives you've considered

Alternative is simple, either extending the Path or doing it inside the view function.

### Environment

* OS: [e.g. Linux / Windows / macOS]: Platform independent
* FastAPI Version [e.g. 0.3.0]: 0.65.2

**Comment by ArcLightSlavik** [comment]
Could you provide an alternatives / how it looks right now?

If there is a way to implement it internally without having to use lambda it would be cool, but currently I disagree that it's readable.

**Comment by Niraj-Kamdar** [comment]
Alternative would be to just do it inside the view function
```python
from fastapi import FastAPI

app = FastAPI()


@app.get("/{animal_name}")
def get_animal(animal_name: str):
    animal_name = animal_name.lower()
    return get_animal_from_db(animal_name)
```

But I think it would be better to decouple pre-processing logic from the business logic.

I just used lambda out of habit but it can be any function. Here, one can use `str.lower` directly.
```python
from fastapi import FastAPI

app = FastAPI()


@app.get("/{animal_name}")
def get_animal(animal_name: str = Path(transform=str.lower):
    animal_name = animal_name.lower()
    return get_animal_from_db(animal_name)
```

**Comment by Niraj-Kamdar** [comment]
This pattern is similar to `argparser` which also allow transforming input object to desired object. 
Here's how `argparser` has implemented this functionality:

```python
class StringToListAction(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        values = list(map(lambda val: val.strip(), values.split(",")))
        setattr(namespace, self.dest, values)
        
group.add_argument(
        "-s",
        "--skips",
        dest="skips",
        action=StringToListAction,
        type=str
    )
```

**Comment by n8sty** [comment]
You should be able to use Pydantic for this, see the `to_lower` parameter of the `constr` type: https://pydantic-docs.helpmanual.io/usage/types/#arguments-to-constr. Alternatively, if you have more complicated input processing that you need to do, that can be extracted into a dependency to keep your view function simpler.

**Comment by Niraj-Kamdar** [comment]
I don't actually need to convert to lower it's just the example. I may need to split, join, etc.
Yeah Dependency idea looks good.

**Comment by Niraj-Kamdar** [comment]
Using pydantic models as dependencies seems to solve the problem. closing it.

**Comment by YuriiMotov** [comment, marked as answer]
There is an easier way. 
You can also annotate your type with `AfterValidator`:

```python
from typing import Annotated, TypeAlias
from fastapi import FastAPI
from pydantic import AfterValidator

app = FastAPI()


MyStr: TypeAlias = Annotated[str, AfterValidator(lambda x: x.lower())]


@app.get("/{animal_name}")
def get_animal(animal_name: MyStr):
    return animal_name
```
